src/lib/contracts.ts

```diff
@@ -13,13 +13,21 @@ import supervisorVersion = require('./supervisor-version');
 export { ContractObject };
 
 export interface ServiceContracts {
-	[serviceName: string]: ContractObject;
+	[serviceName: string]: { contract?: ContractObject; optional: boolean };
 }
 
 export async function containerContractsFulfilled(
 	serviceContracts: ServiceContracts,
-): Promise<{ valid: boolean; unmetServices: string[] }> {
-	const containers = _.values(serviceContracts);
+): Promise<{
+	valid: boolean;
+	unmetServices: string[];
+	fulfilledServices: string[];
+}> {
+	const containers = _(serviceContracts)
+		.values()
+		.map('contract')
+		.filter(c => c != null)
```

Comment:
```suggestion
		.compact()
```
(and with this I think you can drop some `!`s below)

Reply:
nice! (the `!` dropping)

Reply:
The other way to do it (eg with native filter) is with a type guard, ie
```ts
arr.filter((x): x is NonNullable<typeof x> => x != null)